rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf

```diff
@@ -1013,6 +1013,38 @@ SecRule REQUEST_METHOD "!@rx ^(?:GET|HEAD|PROPFIND|OPTIONS)$" \
             setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}',\
             setvar:'tx.%{rule.id}-OWASP_CRS/POLICY/CONTENT_TYPE_NOT_ALLOWED-%{matched_var_name}=%{matched_var}'"
 
+#
+# Restrict charset parameter within the content-type header
+#
+SecRule REQUEST_HEADERS:Content-Type "@rx charset\s*=\s*([^;\s]+)" \
+    "id:920480,\
+    phase:1,\
+    block,\
+    t:none,t:lowercase,\
+    msg:'Request content type charset is not allowed by policy',\
+    logdata:'%{matched_var}',\
+    tag:'application-multi',\
+    tag:'language-multi',\
+    tag:'platform-multi',\
+    tag:'attack-protocol',\
+    tag:'OWASP_CRS/PROTOCOL_VIOLATION/CONTENT_TYPE_CHARSET',\
+    tag:'WASCTC/WASC-20',\
+    tag:'OWASP_TOP_10/A1',\
+    tag:'OWASP_AppSensor/EE2',\
+    tag:'PCI/12.1',\
+    rev:1,\
+    ver:'OWASP_CRS/3.0.0',\
+    severity:'CRITICAL',\
+    capture,\
+    chain"
+    SecRule TX:1 "!@rx ^%{tx.allowed_request_content_type_charset}$" \
+        "t:none,\
+        ctl:forceRequestBodyVariable=On,\
```

Comment:
I've taken this from 920420. I assumed it was mean to log the request body, so it could later be inspected.

Reply:
Without having checked the code completely, but if I remember correctly this is a none-op in v3. Together with `ctl:auditEngine`, `drop` and `expirevar`

Reply:
Thank you for the hint. Do you mean the `ctl:forceRequestBodyVariable=On` is a none-op?

Reply:
@dune73  As far as I understand the v3 code, yes. 

The parser maps this action in line https://github.com/SpiderLabs/ModSecurity/blob/eed6b5f86d00ce7d6bf5e14d6f03647bafd6251d/src/parser/seclang-parser.yy#L2560 to a generic `action::Action` object instead of a specific action subclass. This generic `action::Action` evaluates to `true` in line https://github.com/SpiderLabs/ModSecurity/blob/eed6b5f86d00ce7d6bf5e14d6f03647bafd6251d/src/actions/action.cc#L55 without doing something. If I don't miss anything here, this action does nothing.

It's the same for the actions `drop`, `expirevar` an `ctl:auditEngine`.

Maybe it's worth to open a ticket on ModSecurity v3 and ask for clarification about the indented semantic of `ctl:forceRequestBodyVariable` action. Because I also do not find a description in the reference manual either.